### DISCUSSION TITLE
How to apply the recent librespot fixes in mopidy-spotify?

### DISCUSSION THREAD

**Discussion by The-Judge** [mopidy/mopidy-spotify#405] [Q&A] [answered]
#394 Describes a workaround/fix for the long-term issues with playing spotify resources, but seems not to work with [MiczFlor/RPi-Jukebox-RFID](https://github.com/MiczFlor/RPi-Jukebox-RFID):

I am very sorry to ask how to do all of this again, but I consider myself not being a complete IT-moron but yet, I can't get this to work, unfortunately.
I installed the Phoniebox v2 just as the docs say. No visible error experienced.

Since the librespot compile was missing these components, I then installed:
* `cmake`
* `libclang-dev`

Next, I Installed Rust and Cargo (defaults applied): `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh`).

Using librespot commit `3781a089a69ce9883a299dfd191d90c9a5348819` (which has the [fix from #1309](https://github.com/librespot-org/librespot/pull/1309)), I then executed the non-spotify-connect variant: `cargo run --no-default-features -- --cache . --enable-oauth` and successfully created the `credentials.json` file.
Next, I stopped the cargo process (ctrl+c), created the path `$HOME/.local/share/mopidy/spotify/credentials-cache` and copied `credentials.json` there.

Finally, I restarted the Pi.
I heard the Phoniebox-chime and navigated to the Web-UI. Clicking on "Spotify", Iris asks me for that Client's Identifier, which I set and I'm lead to the Iris UI then.

Navigating to `Settings > Spotify` looks as if I'm not logged in yet and using the search causes JSON errors:

![CleanShot 2024-10-06 at 16 55 10](https://github.com/user-attachments/assets/f5f32ee0-104e-44c4-ac65-4dcad68dd2d3)

What am I missing?
I don't even know which logs to provide ...

Doesn't that new librespot version need to be installed in addition to a good `credentials.json` file or something?

**Comment by kingosticks** [comment]
The librespot stuff is only relevent to playback but you're not even getting that far. What version of Mopidy-Spotify are you using, the latest pre-release? Iris has some logs, check those as well as the normal Mopidy logs.

**Comment by troyxmccall** [comment]
@The-Judge your JSON stuff looks like you need to re-auth here 

https://mopidy.com/ext/spotify/

and update your mopidy config with

```
[spotify]
client_id = The config value will appear here.
client_secret = The config value will appear here.
```

**Reply by kingosticks** [reply]
Also note there's new releases of Mopidy-Spotify and gst-plugin-spotify that fix the recent auth issues so you don't need to mess around with credentials.json yourself.

**Comment by The-Judge** [comment]
Sorry guys, I still don't get the proper steps here ...
I just started all over again following the [Version 2 Setup instructions for RaspberryPi OS Bullseye](https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/INSTALL) (`Raspberry Pi OS (Legacy, 32-bit) Lite` in Raspberry Pi Imager).

I then installed the whole thing using the `development` branch, since I want to have these newest changes to the Spotify Auth:

`cd; rm install-jukebox.sh; wget https://raw.githubusercontent.com/MiczFlor/RPi-Jukebox-RFID/develop/scripts/installscripts/install-jukebox.sh; chmod +x install-jukebox.sh; GIT_BRANCH=develop bash ./install-jukebox.sh`

With the following inputs to the Install Script:

```
################################################
Interactive mode: true
GIT_BRANCH develop
GIT_URL https://github.com/MiczFlor/RPi-Jukebox-RFID.git
Current User: pi
User home dir: /home/pi
Used Raspberry Pi OS: bullseye
################################################
WIFIconfig="NO"
WIFIcountryCode="DE"
EXISTINGuse="NO"
AUTOHOTSPOTconfig="NO"
WIFIconfig="NO"
WIFIcountryCode="DE"
EXISTINGuse="NO"
AUTOHOTSPOTconfig="NO"
AUDIOiFace="DSP Program"
SPOTinstall="YES"
DIRaudioFolders="/home/pi/RPi-Jukebox-RFID/shared/audiofolders"
GPIOconfig="YES"
################################################
```

For Spotify, I entered all 4 components:
1. Username
2. Password
3. client_id
4. client_secret

With 3 and 4 freshly created at https://mopidy.com/ext/spotify/

This runs like a charm, with no visible issues and results in the following:

- Phoniebox boots up with it's chime sound
- `speaker-test -c2` as user `pi` works with both channels
- Web UI is reachable
- Clicking the "Spotify" Icon in the Top Menu opens Iris at http://<MY_IP>:6680/iris/
  - In Iris, navigating to `Settings` says for Spotify "Authorized" and the Tab has the Spotify Profile picture.
  - Searching for anything in Iris displays "Searching Spotify" in the lower right corner, but no results are shown.
  - Similar to that, no playlists or artists can be loaded.
- Uploading and playing MP3 files works great.

> The librespot stuff is only relevent to playback but you're not even getting that far. What version of Mopidy-Spotify are you using, the latest pre-release? Iris has some logs, check those as well as the normal Mopidy logs.

I'm not too sure how to find these infos ...
There's a Debian Package installed named `mopidy`; that is version `3.4.2-2`.
Looking at the Python packages the `pi` user has, this is the list:
```
pi@phoniebox:~ $ python3 -m pip freeze
attrs==24.2.0
Brotli==1.1.0
certifi==2024.8.30
cffi==1.14.5
chardet==4.0.0
charset-normalizer==3.3.2
colorzero==1.1
coverage==7.6.1
cryptography==3.3.2
distro==1.5.0
dnspython==2.0.0
evdev==1.7.1
gpg===1.14.0-unknown
gpiozero==1.6.2
idna==3.10
iniconfig==2.0.0
Markdown==3.3.4
mock==5.1.0
Mopidy==3.4.2
Mopidy-Iris==3.69.3
Mopidy-Local==3.2.1
Mopidy-MPD==3.3.0
Mopidy-Spotify==5.0.0a2
mutagen==1.47.0
numpy==1.19.5
packaging==24.1
pi-rc522==2.3.0
picamera2==0.3.12
pidng==4.0.9
piexif==1.1.3
Pillow==8.1.2
pluggy==1.5.0
ply==3.11
py==1.11.0
pycparser==2.20
pycryptodomex==3.21.0
Pygments==2.7.1
PyGObject==3.38.0
pykka==4.0.2
pyserial==3.5
pytest==6.2.5
pytest-cov==5.0.0
pytest-pythonpath==0.7.4
python-apt==2.2.1
python-mpd2==3.1.1
python-prctl==1.7
PyYAML==5.3.1
requests==2.32.3
requests-toolbelt==0.9.1
RPi.GPIO==0.7.1
simplejpeg==1.6.4
six==1.16.0
SPI-Py @ git+https://github.com/lthiery/SPI-Py.git@b76196d80d9932467998b0163f76c38fb8df45a9
spidev==3.6
ssh-import-id==5.10
toml==0.10.2
tomli==2.0.2
tornado==6.4.1
typing-extensions==4.12.2
uritools==4.0.3
urllib3==2.2.3
v4l2-python3==0.3.2
websockets==13.1
yt-dlp==2024.10.7
pi@phoniebox:~ $
```

So, for what I understand, everything but Spotify is working.
Now what - what are the next steps required to make this work, too? Spotify is by far the most important element for my Kids.

**Reply by kingosticks** [reply, marked as answer]
Check your Mopidy logs with `sudo journalctl -u mopidy`. I think Iris logs are accessible somewhere in its settings UI?

> since I want to have these newest changes to the Spotify Auth

In which case you *must*:

1. Download latest gst-plugin-spotify_0.14.0.alpha.1-1_armhf.deb from [here](https://github.com/kingosticks/gst-plugins-rs-build/releases/download/gst-plugin-spotify_0.14.0-alpha.1-1/gst-plugin-spotify_0.14.0.alpha.1-1_armhf.deb)
2. Install it: `sudo dpkg -i gst-plugin-spotify_0.14.0.alpha.1-1_armhf.deb`
3. Download and install latest Mopidy-Spotify: `sudo python3 -m pip install --break-system-packages Mopidy-Spotify==5.0.0a3`
4. Restart Mopidy `sudo systemctl restart mopidy`.

If you manually created any credentials.json files, remove them and redo step 4.

**Reply by The-Judge** [reply]
There seems to be no such switch to the pip in Bullseye:

```
root@phoniebox:~# python3 -m pip install Mopidy-Spotify==5.0.0a3 --break-system-packages

Usage:
  /usr/bin/python3 -m pip install [options] <requirement specifier> [package-index-options] ...
  /usr/bin/python3 -m pip install [options] -r <requirements file> [package-index-options] ...
  /usr/bin/python3 -m pip install [options] [-e] <vcs project url> ...
  /usr/bin/python3 -m pip install [options] [-e] <local project path> ...
  /usr/bin/python3 -m pip install [options] <archive url/path> ...

no such option: --break-system-packages
root@phoniebox:~#
```

Anyways: It worked without that switch.

Unfortunately, I still can't do anything with Spotify in Iris.

Clicking on "Refresh Token" in Iris also does not work; it results in an infinite spinner animation.

So, in general: Iris seems still broken.
But the good News is, that these steps fix the "Normal" Spotify Playback for me! Thanks a lot! 👯

**Reply by kingosticks** [reply]
> There seems to be no such switch to the pip in Bullseye:

[`python3 -m pip install --upgrade pip`](https://pip.pypa.io/en/stable/installation/#upgrading-pip)

I still think Iris logs are accessible somewhere in its settings UI?

**Comment by dnj0104** [comment]
Hello , thanks for fix the spotify issue, now I just have one question how do i install the new alpha release of mopidy-spotify, if i follow the install steps at readme and run
`sudo python3 -m pip install --break-system-packages Mopidy-Spotify==5.0.0a3`

I get this:

`Usage:   
  /usr/bin/python3 -m pip install [options] <requirement specifier> [package-index-options] ...
  /usr/bin/python3 -m pip install [options] -r <requirements file> [package-index-options] ...
  /usr/bin/python3 -m pip install [options] [-e] <vcs project url> ...
  /usr/bin/python3 -m pip install [options] [-e] <local project path> ...
  /usr/bin/python3 -m pip install [options] <archive url/path> ...
no such option: --break-system-packages`

google says that ` --break-system-packages` does not work anymore kind of. So i went further and downloaded the v5.0.0a3.zip from here and unzipped but now I'm not sure if that was the right thing to do, so I Just ask what should I do now ? 

and sidenote gst-plugin i believe is correct installed:

Plugin Details:
  Name                     spotify
  Description              GStreamer Spotify Plugin
  Filename                 /usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgstspotify.so
  Version                  0.14.0-alpha.1-335dbdd
  License                  MPL
  Source module            gst-plugin-spotify
  Source release date      2024-10-06
  Binary package           gst-plugin-spotify
  Origin URL               https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs

  spotifyaudiosrc: Spotify source

  1 features:
  +-- 1 elements

**Reply by kingosticks** [reply]
I'd be surprised if that option was removed, maybe your version of pip is too old. Just omit that option then the rest of the command is still valid.

**Reply by dnj0104** [reply]
I'm running python3.9. 2 ,  yes it was an old pip version , thank you !

**Comment by dnj0104** [comment]
some reflections or summary how I got mopidy-spotify working now after a couple of hiccups 

here is my way how I get it to work on Debian Bookworm latest:
follow kingosticks instructions in readme: 

1 .Download latest gst-plugin-spotify_0.14.0.alpha.1-1_armhf.deb
Install it: sudo dpkg -i gst-plugin-spotify_0.14.0.alpha.1-1_armhf.deb

2. Download and install latest Mopidy-Spotify: sudo python3 -m pip install --break-system-packages Mopidy-Spotify==5.0.0a3
(in this step I got a hiccup - my pip was old,  solution update pip. Run install command again).

Restart mopidy I have mopidy as service so i run `sudo systemctl restart mopidy`. 

I then tried to start a song from spotify via Iris- but nothing happend, i did check my logs easiest way to check latest logs is just to
punch in _IF_ running mopidy as service - in terminal `sudo systemctl status mopidy` and there I find messages bad credentials.

bad luck for me - I tried to renew my credentials . but no luck , so then i remember I have read about the credentials.

In my case I have the spotify credentials in `/var/lib/mopidy/spotify/credentials-cache` maybe you cant open the credentials-cache directory as me so I had to enter it as root, sudo didnt work ,  proceed to `sudo nano credentials.json` here I noticed some stuff , user was not correct, i tried to change to the "other usernamen,  user was set as MrJones BUT if you login to your spotify account on web you can see that your user also have another username, I changed from MrJones to the other username its just a bunch of numbers and letters.  save and exit nano, exit your root user, restart mopidy, nothing changed , could not play songs from spotify. 

I got slightly aggrivated so I just got in to `/var/lib/mopidy/spotify/credentials-cache` and punch in command as root `rm credentials.json`  exit root 

restart mopidy again `sudo systemctl restart mopidy` and boom playing songs from spotify worked again! 

but I have some playlist that mopidy-spotify wont play, I really dont know why yet but probably it will get loosen up soon.

**Reply by kingosticks** [reply]
I did say...

> If you manually created any credentials.json files, remove them and redo step 4.

**Reply by dnj0104** [reply]
yes! 

I know , but the thing is here that I have not what i know created the credentials.json by myself and moved it to /var/lib/mopidy/spotify/credentials-cache .  Something else did that.  
So I believed that, that step was not necessary for my setup, but it turned out it was a necessary step 

Thanks again all involved devs for the good solution.

**Comment by The-Judge** [comment]
Thank you, @kingosticks and @troyxmccall very, very much! After more than a year, my two Phonieboxes work again and I look into bright children eyes :D

**Comment by Yo55arian** [comment]
Echoing the thanks to @kingosticks and @troyxmccall . Really appreciate all your time and effort in addressing this.

I'm hoping I haven't made a noob mistake, but I've been lurking and keeping an eye on developments here and decided to try the fix above  only to hit an issue.

I've followed the advice from @kingosticks above (https://github.com/mopidy/mopidy-spotify/discussions/405#discussioncomment-10881114) and after completing those steps:

- running `dpkg -l` indicates I have `gst-plugin-spotify:armhf 0.14.0~alpha.1-1` installed.
- similarly, running `python3 -m pip freeze` indicates that I have `Mopidy-Spotify==5.0.0a3` installed. Worth noting that I also encountered the problem with `--break-system-packages` even after updating pip.
- I haven't created a credentials.json file and don't seem to have anything in `/var/lib/mopidy/spotify/credentials-cache` as was reported by @dnj0104 
- I have updated the client_id and client_secret in `/etc/mopidy/mopidy.conf` based on details freshly obtained from https://mopidy.com/ext/spotify/
- I have also removed the username and password lines from `/etc/mopidy/mopidy.conf` so it now reads:
```
[spotify]
enabled = true
client_id = XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
client_secret = YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY
```

Unfortunately I am still not able to play Spotify tracks. The track information loads, but if a try to play a Spotify track using the webapp it tells me the track is playing but nothing happens. I can play MP3 files with no issue, that all seems to be working fine.

Therefore I run `sudo journalctl -u mopidy` to see what is going on, and I can see the following warnings and error messages which seem to point at an issue with gstreamer (messages truncated to what I _think_ seem relevant, but can upload full logs if helpful):

```
Oct 11 23:47:32 raspberrypi mopidy[4814]: WARNING  [MainThread] mopidy_mpd.network Rejected connection from [::ffff:127.0.>
```

```
Oct 12 22:25:31 raspberrypi mopidy[1706]: Traceback (most recent call last):
Oct 12 22:25:31 raspberrypi mopidy[1706]:   File "/usr/local/lib/python3.9/dist-packages/mopidy/audio/actor.py", line 577, in _on_source_setup
Oct 12 22:25:31 raspberrypi mopidy[1706]:     self._source_setup_callback(source)
Oct 12 22:25:31 raspberrypi mopidy[1706]:   File "/usr/local/lib/python3.9/dist-packages/mopidy_spotify/backend.py", line 50, in on_source_setup
Oct 12 22:25:31 raspberrypi mopidy[1706]:     source.set_property("access-token", self.backend._web_client.token())
Oct 12 22:25:31 raspberrypi mopidy[1706]: TypeError: object of type `GstSpotifyAudioSrc' does not have property `access-token'
Oct 12 22:25:31 raspberrypi mopidy[1706]: ERROR    [MainThread] mopidy.audio.gst GStreamer error: Could not get/set settings from/on resource.
```

```
Oct 12 22:27:51 raspberrypi mopidy[1761]: WARNING  [FileBackend-3] mopidy.file.library Failed looking up file:///home/pi/RPi-Jukebox-RFID/share>
```

I'm fairly sure it must be something I have overlooked but would very much appreciate any steer you can offer to set me in the right direction.

Thanks again for the hours already expended on this.

**Reply by kingosticks** [reply]
> TypeError: object of type `GstSpotifyAudioSrc' does not have property `access-token'

Indicates you don't have the correct version of gst-plugin-spotify. What version does `gst-inspect-1.0 spotifyaudiosrc | grep Version` show?

**Reply by Yo55arian** [reply]
Strange, here's what I get:
```
No such element or plugin 'spotifyaudiosrc'
```
Which clearly doesn't seem right.

Here is the output for gst or gstreamer from `dpkg -l`:
```
ii  gst-plugin-spotify:armhf               0.14.0~alpha.1-1                   armhf        GStreamer Spotify Plugin
ii  gstreamer1.0-alsa:arm64                1.18.4-2+deb11u2                   arm64        GStreamer plugin for ALSA
ii  gstreamer1.0-gl:arm64                  1.18.4-2+deb11u2                   arm64        GStreamer plugins for GL
ii  gstreamer1.0-gtk3:arm64                1.18.4-2+deb11u2+rpt1              arm64        GStreamer plugin for GTK+3
ii  gstreamer1.0-libav:arm64               1.18.4-3                           arm64        ffmpeg plugin for GStreamer
ii  gstreamer1.0-plugins-bad:arm64         1.18.4-3+deb11u4                   arm64        GStreamer plugins from the "bad" set
ii  gstreamer1.0-plugins-base:arm64        1.18.4-2+deb11u2                   arm64        GStreamer plugins from the "base" set
ii  gstreamer1.0-plugins-good:arm64        1.18.4-2+deb11u2+rpt1              arm64        GStreamer plugins from the "good" set
ii  gstreamer1.0-plugins-ugly:arm64        1.18.4-2+deb11u1                   arm64        GStreamer plugins from the "ugly" set
ii  gstreamer1.0-pulseaudio:arm64          1.18.4-2+deb11u2+rpt1              arm64        GStreamer plugin for PulseAudio (transitional package)
ii  gstreamer1.0-qt5:arm64                 1.18.4-2+deb11u2+rpt1              arm64        GStreamer plugin for Qt5
ii  gstreamer1.0-tools                     1.18.4-2.1                         arm64        Tools for use with GStreamer
ii  gstreamer1.0-x:arm64                   1.18.4-2+deb11u2                   arm64        GStreamer plugins for X11 and Pango
```

**Reply by kingosticks** [reply]
It looks like you accidentally installed the armhf version.

**Reply by Yo55arian** [reply]
🤦

Well at least an easy fix! Thank you. 

Have now installed the 64bit version and gst-inspect is confirming it's running. Will test if Spotify audio works for certain a little later and confirm.

**Reply by Yo55arian** [reply]
Success! Sorry for the numpty question but it now seems to be working 100%. Great stuff.

**Reply by MrTolchock** [reply]
I had a perfectly well running Mopidy-Spotify, before the token changes came. I then followed your new instructions. Mopidy-Spotify seems to work fine, and Mopidy is actually connecting to Spotify and loading the playlists. However, I don't hear any sound.

Any idea why this could be or how I could try to fix it? Maybe something noteworthy (not sure if relevant): I did the rust stuff for the gst-plugin on another Raspberry Pi with the same OS, because my Pi Zero is too slow for it. The approach worked well for the last gst-plugin.

I have installed the following versions:

- gst-plugin: 0.14.0-alpha.1-5f05516
- Mopidy: Version: 3.4.2
- Mopidy-Spotify: Version: 5.0.0a3

Many thanks for your help, it is much appreciated!

**Comment by MrTolchock** [comment]
Hi there, it is still me, and I just cannot get Mopidy-Spotify to work again. When I try to play a Spotify track, Mopidy gives the following error: 

_TypeError: object of type `GstSpotifyAudioSrc' does not have property `access-token'_

I use Mopidy-Spotify 5.0.0a3 and gst-plugin 0.14.0-alpha.1-5f05516. What can I try to make it work again?

**Reply by kingosticks** [reply]
The Gstreamer Spotify src plugin has required the access-token property for many months. The two latest releases https://github.com/kingosticks/gst-plugins-rs-build support it.

**Reply by MrTolchock** [reply]
Thank you very much! I always failed to cross-compile the plugin, and now realised that I don't have to (I didn't know about .deb files, sorry).

**Reply by kingosticks** [reply]
No worries. Glad it helped.
